project/Build.scala

```diff
@@ -411,37 +417,126 @@ object Build extends sbt.Build {
     settings = commonSettings
   ) dependsOn (wordcount, complexdag, sol, fsio, examples_kafka, distributedshell, stockcrawler, transport)
   
-  lazy val services = Project(
-    id = "gearpump-services",
-    base = file("services"),
-    settings = commonSettings  ++ 
-      Seq(
-        libraryDependencies ++= Seq(
-          "io.spray" %%  "spray-testkit"   % sprayVersion % "test",
-          "io.spray" %%  "spray-httpx"     % sprayVersion,
-          "io.spray" %%  "spray-client"    % sprayVersion,
-          "io.spray" %%  "spray-json"    % sprayJsonVersion,
-          "com.wandoulabs.akka" %% "spray-websocket" % sprayWebSocketsVersion
-            exclude("com.typesafe.akka", "akka-actor_2.11"),
-          "org.scalatest" %% "scalatest" % scalaTestVersion % "test",
-          "org.webjars" % "angularjs" % "1.4.3", // We stick on 1.3.x until angular-ui-select works
-          "org.webjars" % "bootstrap" % "3.3.5",
-          "com.lihaoyi" %% "upickle" % upickleVersion,
-          "org.webjars" % "d3js" % "3.5.5",
-          "org.webjars" % "momentjs" % "2.10.3",
-          "org.webjars.bower" % "angular-loading-bar" % "0.8.0",
-          "org.webjars.bower" % "angular-smart-table" % "2.1.1",
-          "org.webjars.bower" % "angular-motion" % "0.4.2",
-          "org.webjars.bower" % "bootstrap-additions" % "0.3.1",
-          "org.webjars.bower" % "angular-strap" % "2.3.1",
-          "org.webjars.bower" % "angular-ui-select" % "0.12.0",
-          "org.webjars" % "select2" % "3.5.2",
-          "org.webjars" % "select2-bootstrap-css" % "1.4.6",
-          "org.webjars.bower" % "ng-file-upload" % "5.0.9",
-          "org.webjars.bower" % "vis" % "4.5.1"
-        ).map(_.exclude("org.scalamacros", "quasiquotes_2.10")).map(_.exclude("org.scalamacros", "quasiquotes_2.10.3"))
-      )
-  ) dependsOn(streaming % "test->test;compile->compile", daemon % "test->test;compile->compile")
+  lazy val copySharedResources = Def.task {
+    ConsoleLogger().info("copying shared resources")
+//    val sharedMessages = "src/main/scala/org/apache/gearpump/shared/Messages.scala"
+//    IO.copyFile(core.base / sharedMessages, file("services/js") / sharedMessages)
+  }
+
+  lazy val services = Project(id = "gearpump-services", base = file("services")).
+    settings(commonSettings: _*).
+    aggregate(servicesjs, servicesjvm).
+    dependsOn(streaming % "test->test;compile->compile", daemon % "test->test;compile->compile")
+
+  lazy val servicesjvm = Project(id = "gearpump-services-jvm", base = file("services/jvm")).
+    settings(jvmSettings : _*).
+    dependsOn(streaming % "test->test;compile->compile", daemon % "test->test;compile->compile")
+
+  lazy val servicesjs = Project(id = "gearpump-services-js", base = file("services/js")).
+    enablePlugins(ScalaJSPlugin).
+    settings(commonSettings: _*).
+    settings(jsSettings : _*)
+
+  def findDirs(currentDir: String): Array[File] = {
```

Comment:
@clockfly 
I'll update to use baseDirectory.value / ..
@huafengw 
CopyJSArtifacts needs to be called as part of fastOptJS actions so that ~fastOptJS works. 
```
    fastOptJS in Compile := {
      val originalResult = (fastOptJS in Compile).value
      copyJSArtifactsToOutput
      originalResult
    },
```
This mean copying specific files to output dir as well as modifying the .map file so scala debug works correctly in a browser. These actions will be triggered when any file is updated.

@manuzhang 
I can put some of the utility functions in a file similar to what has been done in PhantomJS2Env.scala.